### PR TITLE
Allow independent methods to have the same parameter name

### LINKED ISSUE #45492: Parameter name is suggested with number suffix while it should be plain
**Version Used**: 16.7-preview3.1

**Steps to Reproduce**:

1.  In class, create method with CancellationToken parameter and name `cancellationToken`
2.  Next, create another method also with CancellationToken, observe suggested name
image
image

**Expected Behavior**:
IDE should suggest `cancellationToken` parameter name
IDE should not add suffix number to parameter name. suffix should be added only if we are using same type in same method.

**Actual Behavior**:
IDE suggests `cancellationToken1` parameter name.

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #45492

### CODE REVIEW

**Comment by miqm** [comment]
@sharwell What in case of local methods? In C#8 we can hide existing parameter, but in netcore2.1 suggested name can breake up build.

**Comment by sharwell** [comment]
@miqm I believe that's an existing bug, so I wasn't trying to fix it. Is there a specific case that worked in 16.6 but isn't working anymore?

**Comment by miqm** [comment]
@sharwell - I think I didn't make myself clear enough. The full test case, that IMHO should be are as follows
```
{
var markup = @"
public class DbContext { }
public class C
{
    void Goo(DbContext context) {
        void InnerGoo(DbContext $$) { }
    }
}
";
await VerifyItemExistsAsync(markup, "dbContext", glyph: (int) Glyph.Parameter);
await VerifyItemExistsAsync(markup, "db", glyph: (int) Glyph.Parameter);
await VerifyItemExistsAsync(markup, "context1", glyph: (int)Glyph.Parameter);
}
```
```
{
var markup = @"
public class DbContext { }
public class C
{
    void Goo() {
        DbContext context;
        void InnerGoo(DbContext $$) { }
    }
}
";
await VerifyItemExistsAsync(markup, "dbContext", glyph: (int) Glyph.Parameter);
await VerifyItemExistsAsync(markup, "db", glyph: (int) Glyph.Parameter);
await VerifyItemExistsAsync(markup, "context1", glyph: (int)Glyph.Parameter);
}
```
```
{
var markup = @"
public class DbContext { }
public class C
{
    DbContext dbContext;
    void Goo(DbContext context) {
        void InnerGoo(DbContext $$) { }
    }
}
";
await VerifyItemExistsAsync(markup, "dbContext", glyph: (int) Glyph.Parameter);
await VerifyItemExistsAsync(markup, "db", glyph: (int) Glyph.Parameter);
await VerifyItemExistsAsync(markup, "context1", glyph: (int)Glyph.Parameter);
}
```
I verified those test cases on 16.7-preview2 and they all work. I'm not sure if they will work with your code. Stopping search on ParameterList syntax will take into account only items on current parameter list (inner method's list) and will not take into account the enclosing block.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs [hunk 1]
```diff
@@ -182,6 +182,9 @@ public SyntaxNode GetDefaultOfParameter(SyntaxNode node)
         public SyntaxNode GetParameterList(SyntaxNode node)
             => node.GetParameterList();
 
+        public bool IsParameterList(SyntaxNode node)
```
Comment (miqm) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs:185]:
```
        public bool IsParameterScope(SyntaxNode node)
            => node.IsKind(SyntaxKind.Block, SyntaxKind.SwitchSection);
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Services/SyntaxFacts/ISyntaxFacts.cs [hunk 2]
```diff
@@ -208,6 +208,7 @@ internal partial interface ISyntaxFacts
         SyntaxToken? GetNameOfParameter(SyntaxNode node);
         SyntaxNode GetDefaultOfParameter(SyntaxNode node);
         SyntaxNode GetParameterList(SyntaxNode node);
+        bool IsParameterList(SyntaxNode node);
```
Comment (miqm) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Services/SyntaxFacts/ISyntaxFacts.cs:211]: `bool IsParameterScope(SyntaxNode node);`

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/SyntaxFacts/VisualBasicSyntaxFacts.vb [hunk 3]
```diff
@@ -202,6 +202,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.LanguageServices
             Return node.GetParameterList()
         End Function
 
+        Public Function IsParameterList(node As SyntaxNode) As Boolean Implements ISyntaxFacts.IsParameterList
```
Comment (miqm) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/SyntaxFacts/VisualBasicSyntaxFacts.vb:205]:
```
        Public Function IsParameterScope(node As SyntaxNode) As Boolean Implements ISyntaxFacts.IsParameterScope
            Return node.IsKind(SyntaxKind.Block, SyntaxKind.SwitchSection)
        End Function
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/SemanticsFactsService/AbstractSemanticFactsService.cs [hunk 4]
```diff
@@ -58,7 +58,7 @@ internal abstract class AbstractSemanticFactsService
             IEnumerable<string> usedNames, CancellationToken cancellationToken)
         {
             var container = containerOpt ?? location.AncestorsAndSelf().FirstOrDefault(
-                a => SyntaxFacts.IsExecutableBlock(a) || SyntaxFacts.IsMethodBody(a));
+                a => SyntaxFacts.IsExecutableBlock(a) || SyntaxFacts.IsParameterList(a) || SyntaxFacts.IsMethodBody(a));
```
Comment (miqm) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/SemanticsFactsService/AbstractSemanticFactsService.cs:61]:
```
                a => SyntaxFacts.IsParameterScope(a) || SyntaxFacts.IsMethodBody(a));
```

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/SemanticsFactsService/AbstractSemanticFactsService.cs:61]: IsParameterScope isn't something SyntaxFacts should know about.  SyntaxFacts is just a way to structurally ask what type of node something is, in a way that is agnostic to VB/C#.  i.e. if both VB/C# have an InvocationExpressionSyntax, then SyntaxFacts can be used with IsInvocationExpression to ask that question in a uniform way, without worrying about lang. 

It's specifically for syntactic similarity questions, not for semantic questions (like scopes).  Thanks!

Comment (miqm) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/SemanticsFactsService/AbstractSemanticFactsService.cs:61]: @CyrusNajmabadi Ok, understood.

**Comment by miqm** [comment]
> @miqm I believe that's an existing bug, so I wasn't trying to fix it. Is there a specific case that worked in 16.6 but isn't working anymore?

@sharwell This is screenshot from 16.6.2:
image

As you can see, context1 name is suggested here correctly - context is used in the method parameters list. In this solution I think "context" will be suggested, as the lookup for container will stop on first Parameters list - in this case - on the parameters list for the local method.

**Comment by sharwell** [comment]
@miqm I added all your tests locally and they worked without other changes.

**Pull request merged**